File: Data-Structures/Cpp/Misc/BST_orderTraversal.cpp
#include<iostream>
using namespace std;

class Node{
	public:
	int data;
	Node* left;
	Node* right;
	Node* parent;
	
	Node(int D){
		data = D;
	}
	
	~Node(){
		cout<<"Node is deleted"<<endl;
	}
};

class BST{
	public:
		Node* head = NULL;
		
		BST(int s){
			if (s>0){
				cout<<"Enter the head value: ";
				int h;
				cin>>h;
				head = new Node(h);
				Node* first = head;
				for(int i=1; i<s; i++){
					cout<<"Enter the node's Data: ";
					int d;
					cin>>d;
					insertNode(d);
				}
			}
		}
		
		~BST(void){
			cout<<"BST is Deleted"<<endl;
			while(head!=NULL){
				deleteNode(head->data);
			}
		}
		
		bool isNull(){
			return head==NULL;
		}
		
		bool find(int n){
			Node *first = head;
			while(first!=NULL){
				if(first->data == n){
					return true;
				}
				else if(first->data>n){
					first = first->right;
				}
				else{
					first = first->left;
				}
			}
			return false;
		}
		
		Node* findMin(){
			if(head==NULL){
				return NULL;
			}
			Node* first = head;
			while(first->left!=NULL){
				first = first->left;
			}
			return first;
		}
		Node* findMax(){
			if(head==NULL){
				return NULL;
			}
			Node* first = head;
			while(first->right!=NULL){
				first = first->right;
			}
			return first;
		}
		Node* findMin(Node* start){
			Node* first = start;
			while(first->left!=NULL){
				first = first->left;
			}
			return first;
		}
		Node* findMax(Node* start){
			Node* first = start;
			while(first->right!=NULL){
				first = first->right;
			}
			return first;
		}
		
		void insertNode(int d){
			Node* first = head;
			Node* p = NULL;
			while(first!=NULL){
				p = first;
				if(first->data>d){
					first = first->left;
				}else if(first->data<d){
					first = first->right;
				}else{
					cout<<"Node Already Exists!!!"<<endl;
					return;
				}
			}
			first = new Node(d);
			if(p!=NULL){
				first->parent = p;
			}
		}
		
		
		void deleteNodeCase2(Node* first){
			if(first->parent==NULL){
				if(first->left==NULL){
					head = first->right;
					delete first;
				}else if(first->right==NULL){
					head = first->left;
					delete first;
				}
			}else if(first->left==NULL){
				if(first->parent->left->data==first->data){
					Node* sec = first->right;
					first->parent->left = sec;
					delete first;
				}else{
					Node* sec = first->right;
					first->parent->right = sec;
					delete first;
				}
			}else{
				if(first->parent->left->data==first->data){
					Node* sec = first->left;
					first->parent->left = sec;
					delete first;
				}else{
					Node* sec = first->left;
					first->parent->right = sec;
					delete first;
				}
			}
		}
		
		void deleteNodeCase3(Node* first){
			Node* TBR = findMin(first->right);
			int temp = TBR->data;
			deleteNode(TBR->data);
			first->data = temp;
		}
		
		void deleteNode(int d){
			Node* first = head;
			while(first!=NULL){
				if(first->data>d){
					first = first->left;
				}else if(first->data<d){
					first = first->right;
				}else{
					if(first->left==NULL&&first->right==NULL){
						delete first;
					}else if(first->left==NULL||first->right==NULL){
						deleteNodeCase2(first);
					}else{
						deleteNodeCase3(first);
					}
				}
			}
		}
		
		void preOrderTraversal(Node* node){
			if (node == NULL){
		        return;
			}
		    cout<<(node->data);
		    preOrderTraversal(node->left);
		    preOrderTraversal(node->right);
		}
		void postOrderTraversal(Node* node){
			if (node == NULL){
		        return;
			}
		    postOrderTraversal(node->left);
		    postOrderTraversal(node->right);
		    cout<<(node->data);
		}
		void inOrderTraversal (Node* node){
			if (node == NULL){
		        return;
			}
		    inOrderTraversal(node->left);
		    cout<<(node->data);
		    inOrderTraversal(node->right);
		}
		
};



void Q1(){
	cout<<"Enter the length of the BST: ";
	int s;
	cin>> s;
	BST* A = new BST(s);
	cout<<"IsNull: "<<A->isNull()<<endl;
	
	A->inOrderTraversal(A->head);
	A->preOrderTraversal(A->head);
	A->postOrderTraversal(A->head);
//	cout<<"Enter the integer you want to find in BST: ";
//	int f;
//	cin>>f;
//	if(A->find(f)){
//		cout<<"The number lies in the BST..."<<endl;
//	}
//	else{
//		cout<<"The number is not found in the BST..."<<endl;
//	}
	
//	cout<<"Enter the Value to add in the BST: ";
//	int as;
//	cin>> as;
//	A->insertNode(as);

//	cout<<"Enter number to del from the BST: ";
//	int d;
//	cin>>d;
//	A->deleteNode(d);
}

int main(){
	Q1();
}

File: Data-Structures/Python/BST_orderTraversal.py
class Node:
    def __init__(self, d):
        self.data = d
        self.left = None
        self.right = None
        self.parent = None
	
    def __del__(self):
        print("Node is deleted")
	


class BST:
    head = None
    
    def __init__(self, s):
    # BST(int s):
        if s>0:
            self.head = Node(int(input("Enter the head value: ")))
            first = self.head
            for i in range(1, s):
            # for(int i=1 i<s i++):
                # print(
                # int d
                # cin>>d
                self.insertNode(int(input("Enter the node's Data: ")))
            
        
    
    
    def __del__(self):
        print("BST is Deleted")
        
    def isNull(self):
        return self.head==None
    
    
    def find(self, n):
        first = self.head
        while first!=None:
            if first.data == n:
                return True
            
            elif first.data>n:
                first = first.right
            
            else:
                first = first.left
            
        
        return False
    
    def findMin(self, start=head):
        first = start
        while first.left!=None:
            first = first.left
        
        return first
    
    def findMax(self, start=head):
        first = start
        while first.right!=None:
            first = first.right
        
        return first
    
    
    def insertNode(self, d):
        first = self.head
        p = None
        while first!=None:
            p = first
            if first.data>d:
                first = first.left
            elif first.data<d:
                first = first.right
            else:
                print("Node Already Exists!!!")
                return
        first = Node(d)
        if p!=None:
            first.parent = p
            if d<p.data:
                p.left = first
            else:
                p.right = first
        else:
            self.head = first
        
    def deleteNodeCase2(self, first):
        if first.parent==None:
            if first.left==None:
                self.head = first.right
                del first
            elif first.right==None:
                self.head = first.left
                del first
            
        elif first.left==None:
            if first.parent.left.data==first.data:
                sec = first.right
                first.parent.left = sec
                del first
            else:
                sec = first.right
                first.parent.right = sec
                del first
            
        else:
            if first.parent.left.data==first.data:
                sec = first.left
                first.parent.left = sec
                del first
            else:
                sec = first.left
                first.parent.right = sec
                del first
    
    def deleteNodeCase3(self, first):
        TBR = self.findMin(first.right)
        temp = TBR.data
        self.deleteNode(TBR.data)
        first.data = temp
    
    
    def deleteNode(self, d):
        first = self.head
        while first!=None:
            if first.data>d:
                first = first.left
            elif first.data<d:
                first = first.right
            else:
                if first.left==None and first.right==None:
                    del first
                elif first.left==None or first.right==None:
                    self.deleteNodeCase2(first)
                else:
                    self.deleteNodeCase3(first)
                
    def preOrderTraversal(self, node):
        if node == None:
            return
        
        print(node.data, end=", ")
        self.preOrderTraversal(node.left)
        self.preOrderTraversal(node.right)
    
    def postOrderTraversal(self, node):
        if (node == None):
            return
        
        self.postOrderTraversal(node.left)
        self.postOrderTraversal(node.right)
        print(node.data, end=", ")
    
    def inOrderTraversal(self, node):
        if (node == None):
            return
        
        self.inOrderTraversal(node.left)
        print(node.data, end=", ")
        self.inOrderTraversal(node.right)
    
def Q1():
    A = BST(int(input("Enter the length of the BST: ")))
    print("IsNull: ", A.isNull())

    A.inOrderTraversal(A.head)
    print()
    A.postOrderTraversal(A.head)
    print()
    A.preOrderTraversal(A.head)
    print()

if __name__=="__main__":
	Q1()

